torch/utils/mem_tracker.py

```diff
@@ -0,0 +1,242 @@
+        fqn = get_fqn()
+        memory_tracking[fqn]["Before Backward"] = get_current_memory_allocated()
+
+    return f
+
+
+def exit_module_backward(name: str):
+    @torch.no_grad
+    def f(module: nn.Module, grad_input: Any, grad_output: Any):
+        global parents
+        assert parents[-1] == name
+        fqn = get_fqn()
+        memory_tracking[fqn]["After Backward"] = get_current_memory_allocated()
+        parents.pop()
+
+    return f
+
+
+def final_call():
+    fqn = get_fqn()
+    memory_tracking[fqn]["After Backward"] = get_current_memory_allocated()
+    parents.pop()
+
+
+def register_hooks(name: str, module: nn.Module):
+    module.register_forward_pre_hook(enter_module_forward(name))
+    module.register_forward_hook(exit_module_forward(name))
+    module.register_full_backward_pre_hook(enter_module_backward(name))
+    if name == "Root":
+        return
+    module.register_full_backward_hook(exit_module_backward(name))
+
+
+def instrument_module(mod: nn.Module):
+    register_hooks("Root", mod)
+
+    for name, module in mod.named_children():
+        if hasattr(module, "inplace") and getattr(module, "inplace"):
+            setattr(module, "inplace", False)
+        register_hooks(name, module)
+
+    def grad_hook(param: nn.Parameter):
+        global WINFO
+        winfo = WINFO[param.grad.untyped_storage()]
+        assert winfo is not None, "grad tensor not found in WINFO"
+        winfo.reftype = RefType.gradient
+
+    for param in mod.parameters():
+        param.register_post_accumulate_grad_hook(grad_hook)
+        winfo = WINFO[param.untyped_storage()]
+        assert winfo is not None, "param tensor not found in WINFO"
+        winfo.reftype = RefType.parameter
+
+    for buffer in mod.buffers():
+        winfo = WINFO[buffer.untyped_storage()]
+        assert winfo is not None, "buffer not found in WINFO"
+        winfo.reftype = RefType.buffer
+
+
+def instrument_optimizer(optim: torch.optim.Optimizer):
+    def outopt(optimizer: torch.optim.Optimizer, args: Any, kwargs: Any) -> None:
+        global FIRST_OPT_ITER, WINFO, MEMORY_USE
+        if FIRST_OPT_ITER:
+            for param, states in optimizer.state.items():
+                for val in states.values():
+                    if isinstance(val, torch.Tensor):
+                        winfo = WINFO[val.untyped_storage()]
+                        assert winfo is not None, "opt state tensor not found in WINFO"
+                        winfo.reftype = RefType.optstate
+            FIRST_OPT_ITER = False
+
+    optim.register_step_post_hook(outopt)
+
+
+def print_mem_stats(stats: Dict[str, int]):
+    for type, mem in stats.items():
+        print(f"\t{type}: {round(mem/MB, 2)} MBs")
+
+
+def display_mem_stats():
+    for mod in memory_tracking.keys():
+        print(f"Module: ", mod)
+        for k, stats in memory_tracking[mod].items():
+            print(f"{k}")
+            print_mem_stats(stats)
+        print()
+
+
+def experiment():
```

Comment:
We should move this to a test file like the flop_counter.
Also we should check there that:
- The measured memory matched max_allocated to an epsilon
- It does run smoothly in fake mode and gives the same result as in real mode

Reply:
@albanD where is this file?

Reply:
https://github.com/pytorch/pytorch/blob/f9b2ffa7c494cd1a0481149cb4089da4b33e4b66/test/test_flop_counter.py